### PR TITLE
[android] Bump targetSdkVersion to 29.

### PR DESCRIPTION
**Pull request by mczernek**
# STATUS UPDATE

- **This change has been deployed to SDK38 Android builders on August 5th, 14:38 UTC.**
- Android apps built **after that moment** in time will have the **`targetSdkVersion` set to 29**.
- Apps built **before that moment** will **remain on `targetSdkVersion = 28`**.
- You **will need to rebuild your app** after that moment for your AAB/APK to include this change.
- **If you're using `turtle-cli`** to build your standalone apps, to have this change included in your build, please upgrade to version equal or higher than `0.17.1`.

---

# Why

We want to target Android 10, which means Android SDK 29.

# How

* Bump targetSdkVersion to 29. 
* Align permissions module to handle new location permissions behaviour. In older sdks we add `foregroundGranted` and `foregroundScope` to indicate what is the status of foreground permissions, while `scope` and `granted` indicated whether bacgrkound permission location is granted and its scope. In new SDK we change scope to work as on iOS (always or whenInUse). What was previously under `scope` key on Android (`fine` or `coarse`) is moved to `accuracy` field.
* Change way of retrieving location for assets in MediaLibrary
* Opt out from scoped external storage in Android. That would need to be addressed before we switch to targeting Android 11

# Test Plan

Test-suite and ncl.

### CODE REVIEW

**Comment by mczernek** [comment]
I'll need some comments on which of those changes needs to be backported to old ShellApps and to which ones @ide.

My suggestion is to backport permissions related changes to SDK 36 to 38. Those SDK has new permissions architecture on Android and backporting should be relatively easy. That would mean we need to backport MediaLibrary changes there, since while we bump targetSdk current location retrieval crashes. That would also mean we do not bump targetSdkVersion for older SDKs (35 and below). Is that ok?

File: apps/test-suite/tests/Permissions.js [hunk 1]
```diff
@@ -131,5 +131,35 @@ export function test(t) {
...
+
+        if (Platform.OS === 'android') {
+          t.expect(locationPermissionKeys).toContain('android');
+          const androidLocationPermissionKeys = Object.keys(
+            result.permissions[Permissions.LOCATION]['android']
+          );
+          t.expect(androidLocationPermissionKeys).toContain('accuracy');
+        }
+
+        console.log('yszty: ', JSON.stringify(result));
```
Comment (lukmccall) [inline apps/test-suite/tests/Permissions.js:161]:
```suggestion
```
🚨

File: packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt [hunk 2]
```diff
@@ -1,36 +1,50 @@
...
 import org.unimodules.interfaces.permissions.PermissionsStatus
 
 class LocationRequester : PermissionRequester {
-  override fun getAndroidPermissions(): List<String> = listOf(
-      Manifest.permission.ACCESS_FINE_LOCATION,
-      Manifest.permission.ACCESS_COARSE_LOCATION
-  )
+  override fun getAndroidPermissions(): List<String> =
+    if (
+      Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
```
Comment (lukmccall) [inline packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt:21]:
```suggestion
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
```

File: packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt [hunk 3]
```diff
@@ -40,10 +54,20 @@ class LocationRequester : PermissionRequester {
           PermissionsStatus.UNDETERMINED.status
         }
       })
+      scope =
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+          val accessBackgroundLocation = permissionsResponse.getValue(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
+          if (accessBackgroundLocation.status == PermissionsStatus.GRANTED) SCOPE_ALWAYS else SCOPE_IN_USE
```
Comment (lukmccall) [inline packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt:60]:
```suggestion
          if (accessBackgroundLocation.status == PermissionsStatus.GRANTED) {
            SCOPE_ALWAYS 
          } else { 
            SCOPE_IN_USE
          }
```
With brackets, it's more readable ;) cause you can see what returns the `when` statement.

File: packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt [hunk 4]
```diff
@@ -40,10 +54,20 @@ class LocationRequester : PermissionRequester {
...
+        } else if (accessCoarseLocation.status == PermissionsStatus.GRANTED || accessFineLocation.status == PermissionsStatus.GRANTED) {
+          SCOPE_ALWAYS
+        } else SCOPE_NONE
```
Comment (lukmccall) [inline packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt:63]:
```suggestion
        } else { 
          SCOPE_NONE
        }
```
same here

**Review by lukmccall — COMMENTED** [review]
It looks good 💪
I've left some comments. 
Don't forget to update permission types in the `expo-permissions` and `expo-localization` (but first, rebase your branch - I've changed them lately).

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryConstants.java [hunk 5]
```diff
@@ -66,151 +67,149 @@
...
-      {"int", ExifInterface.TAG_DEFAULT_CROP_SIZE},
-      {"int", ExifInterface.TAG_ORF_PREVIEW_IMAGE_START},
-      {"int", ExifInterface.TAG_ORF_PREVIEW_IMAGE_LENGTH},
-      {"int", ExifInterface.TAG_ORF_ASPECT_FRAME},
-      {"int", ExifInterface.TAG_RW2_SENSOR_BOTTOM_BORDER},
-      {"int", ExifInterface.TAG_RW2_SENSOR_LEFT_BORDER},
-      {"int", ExifInterface.TAG_RW2_SENSOR_RIGHT_BORDER},
-      {"int", ExifInterface.TAG_RW2_SENSOR_TOP_BORDER},
-      {"int", ExifInterface.TAG_RW2_ISO},
+    {"string", ExifInterface.TAG_ARTIST},
```
Comment (tsapeta) [inline packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryConstants.java:84]: I think it's a good change, but not really related to this PR so maybe commit such formatting changes to master separately? 😅

File: packages/unimodules-permissions-interface/src/PermissionsInterface.ts [hunk 6]
```diff
@@ -11,4 +11,5 @@ export interface PermissionResponse {
   expires: PermissionExpiration;
   granted: boolean;
   canAskAgain: boolean;
+  scope: 'whenInUse' | 'always';
```
Comment (tsapeta) [inline packages/unimodules-permissions-interface/src/PermissionsInterface.ts:14]: What about `never`? 🤔

File: packages/expo-permissions/CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,8 @@
 
 ### 🛠 Breaking changes
 
+- Add `scope` field in returned value to indicate whether background permissions are granted. Add `android.accuracy` field to determine whether `coarse` or `fine` location permission is granted.([#9446](https://github.com/expo/expo/pull/9446) by [@mczernek](https://github.com/mczernek))
```
Comment (tsapeta) [inline packages/expo-permissions/CHANGELOG.md:7]: Let's move this to `expo-location`'s changelog 😉

**Review by tsapeta — APPROVED** [review]

File: packages/expo-location/CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,8 @@
 
 ### 🛠 Breaking changes
 
+- Add `scope` field in returned value to indicate whether background permissions are granted. Add `android.accuracy` field to determine whether `coarse` or `fine` location permission is granted.([#9446](https://github.com/expo/expo/pull/9446) by [@mczernek](https://github.com/mczernek))
```
Comment (bbarthec) [inline packages/expo-location/CHANGELOG.md:7]: This is not sufficient IMO 🤔 
You've made two changes:
1) You moved `scope` into `UMPermissionsResponse` ⬅ that affects `expo-permissions` and I believe separate `CHANGELOG` entry should be there.
2) You added this `android.accuracy` thing only in `expo-location` and that should be stated here in this package as you've done it.

So please add separate missing `CHANGELOG` entry for `expo-permissions` 🙏

**Review by bbarthec — CHANGES_REQUESTED** [review]
LGTM 👍 
I wish it could be split into two separate PRs, but duh .. 🤷

File: packages/expo-location/CHANGELOG.md (hunk 8 above)
Comment (mczernek) [reply packages/expo-location/CHANGELOG.md]: We've had discussion with @tsapeta - scope was actually used only in Location permission. So, including this in `expo-permissions` changelog under `Breaking changes` would be in fact disinformative - would suggest that anyone using `expo-permissions` should be warned, while only those using `expo-location` should.

File: packages/unimodules-permissions-interface/src/PermissionsInterface.ts (hunk 6 above)
Comment (mczernek) [reply packages/unimodules-permissions-interface/src/PermissionsInterface.ts]: `none` - but still it was missing. Added.

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryConstants.java (hunk 5 above)
Comment (mczernek) [reply packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryConstants.java:82]: It would be very hard now...

**Review by bbarthec — APPROVED** [review]
@mczernek, I see 👍 Therefore let's ship it, shall we? 🎉

File: packages/unimodules-permissions-interface/src/PermissionsInterface.ts [hunk 9]
```diff
@@ -11,4 +11,5 @@ export interface PermissionResponse {
   expires: PermissionExpiration;
   granted: boolean;
   canAskAgain: boolean;
+  scope?: 'whenInUse' | 'always' | 'none';
```
Comment (lukmccall) [inline packages/unimodules-permissions-interface/src/PermissionsInterface.ts:14]: I'm not fun of adding the `scope` key here. I think we should add it to the `Permissions.types.ts` and the `expo-localizations` package.

**Review by lukmccall — CHANGES_REQUESTED** [review]
You forget about the [`MediaLibrary.ts`](https://github.com/expo/expo/blob/04e98becc3e7f8a8a1a99eb7f384da98bc0d3762/packages/expo-media-library/src/MediaLibrary.ts#L15).  `scoped` should be renamed to `accessPrivilages`. Same in the [`ImagePicker.types.ts`](https://github.com/expo/expo/blob/04e98becc3e7f8a8a1a99eb7f384da98bc0d3762/packages/expo-image-picker/src/ImagePicker.types.ts#L7).

File: packages/unimodules-permissions-interface/src/PermissionsInterface.ts (hunk 9 above)
Comment (mczernek) [reply packages/unimodules-permissions-interface/src/PermissionsInterface.ts]: Agrred, partially. Moved to expo-permissions.

Comment (lukmccall) [reply packages/unimodules-permissions-interface/src/PermissionsInterface.ts]: It needs to change this schema too - > https://github.com/expo/expo/blob/04e98becc3e7f8a8a1a99eb7f384da98bc0d3762/packages/expo-location/src/Location.types.ts#L210-L221

**Comment by lukmccall** [comment]
See https://github.com/expo/expo/pull/9446#discussion_r463494022

File: packages/expo-image-picker/ios/EXImagePicker/EXImagePickerCameraRollPermissionRequester.m [hunk 10]
```diff
@@ -39,7 +39,7 @@ - (NSDictionary *)getPermissions
   }
   return @{
     @"status": @(status),
-    @"scope": scope,
+    @"accessPrivilages": scope,
```
Comment (ide) [inline packages/expo-image-picker/ios/EXImagePicker/EXImagePickerCameraRollPermissionRequester.m:42]:
```suggestion
    @"accessPrivileges": scope,
```

This change needs to be made across the entire PR.

**Review by ide — COMMENTED** [review]
- For SDK 38 (back-porting) we need to keep the JS API the same as it was before; we cannot have breaking changes within an SDK version. Developers who have written apps with the SDK 38 Permissions API should be able to just run `expo build:android` without modifying their JS code.
- `accessPrivilages` needs to be renamed `accessPrivileges` everywhere

**Comment by tomasgvivo** [comment]
When will this changes be available for building apps?

**Comment by ide** [comment]
We are working on adding this to SDK 38 builds this week.

**Comment by tomasgvivo** [comment]
@ide Great!
Are there any plans on older SDK versions? like 37 or 36?

**Comment by ide** [comment]
No, we are planning on backporting this only to SDK 38.

**Comment by actuallymentor** [comment]
> We are working on adding this to SDK 38 builds this week.

I see this PR is closed, where can we monitor progress? Right now any newly built apps are rejected by the Play console.

**Comment by amsimoes** [comment]
Guys is it possible to build for level 29 using turtle-cli?

I really need to submit a new app ASAP. Thanks!

**Comment by jeisonng** [comment]
> Guys is it possible to build for level 29 using turtle-cli?
> 
> I really need to submit a new app ASAP. Thanks!

you can run Ejecting to Bare Workflow and change the targetSdk while waiting for the update

**Comment by AdamJNavarro** [comment]
Hey all, we'll still post here when we have updated the builders for SDK38.

**Comment by esdrasevt** [comment]
> No, we are planning on backporting this only to SDK 38.

> My suggestion is to backport permissions related changes to SDK 36 to 38. Those SDK has new permissions architecture on Android and backporting should be relatively easy.

@ide , This quote from @mczernek implies that SDK 36, 37, and 38 will get the backport? Am I misinterpreting it?

**Comment by ide** [comment]
That was just a suggestion. We are backporting this to SDK 38, as it is a non-trivial amount of work and Google's policy affects only new apps for now. Existing apps have until November I believe.

**Comment by sjchmiela** [comment]
Good news, folks! 🎉 

- **This change has been deployed to SDK38 Android builders on August 5th, 14:38 UTC.**
- Android apps built **after that moment** in time will have the **`targetSdkVersion` set to 29**.
- Apps built **before that moment** will **remain on `targetSdkVersion = 28`**.
- You **will need to rebuild your app** after that moment for your AAB/APK to include this change.
- **If you're using `turtle-cli`** to build your standalone apps, to have this change included in your build, please upgrade to version equal or higher than `0.17.1`.

**Comment by polaroi8d** [comment]
Coool. I successfully added a new app. Thanks for the fast fix.

**Comment by Kayskip** [comment]
Thanks! Quick fix, you guys are on to it.
Just had to 'expo upgrade' to the latest version.

**Comment by ajacquierbret** [comment]
Thanks a lot ! Works like a charm !

**Comment by Razzze** [comment]
From 5 august i got Network request failed (device and emulator) on all android devices but not for all API requests, why ?

**Comment by xsats** [comment]
I'm running expo SDK 37 and unable to submit android apks due to this problem. 

Does this mean that as it stands, the only way to submit apps to Google is using SDK 38? So far I have not upgraded to SDK 38 due to breaking changes

**Comment by ide** [comment]
Existing apps in the Google Play Store can continue to use Android targetSdk 28 (up to and including Expo SDK 37) until November.

All new apps for the Google Play Store require Android targetSdk 29 (Expo SDK 38 and newer).

So if you already have your app in the Play Store you have at least two and a half months to update your Expo SDK version to 38 (or 39 by the time November comes around). If the app isn't in the Play Store then you will need to use Expo SDK 38 as of right days ago.

**Comment by xsats** [comment]
> Existing apps in the Google Play Store can continue to use Android targetSdk 28 (up to and including Expo SDK 37) until November.
> 
> All new apps for the Google Play Store require Android targetSdk 29 (Expo SDK 38 and newer).
> 
> So if you already have your app in the Play Store you have at least two and a half months to update your Expo SDK version to 38 (or 39 by the time November comes around). If the app isn't in the Play Store then you will need to use Expo SDK 38 as of right days ago.

Many thanks for the clarification. Our case relates to an as yet unpublished app so guess we better get updated to SDK 38

**Comment by imdanielpiva** [comment]
Hi, everyone, thanks for all of the amazing work you've done!

I just couldn't figure out why some of our apps (running on ```"expo": "^38.0.9"```) are getting this error, as they've been built just today.

Is there any reason why it would happen?

**Comment by Alain1405** [comment]
We have just noticed that our SDK 37 (not ejected) using SDK 37, published before August, is not getting OTA updates on Android 10 devices. 
Is this related to this issue? Is OTA checking that the target device is compatible? If so, do we have any way to specify that OTA updates with SDK37 target Android 10 as well? Or upgrade to SDK >37 is necessary?

**Pull request merged**